Services/Database/classes/Atom/class.ilAtomQuery.php

```diff
@@ -0,0 +1,320 @@
+<?php
+require_once('./Services/Database/exceptions/exception.ilDatabaseException.php');
+
+/**
+ * Class ilAtomQuery
+ *
+ * Use ilAtomQuery to fire Database-Actions which have to be done without beeing influenced by other queries or which can influence other queries as
+ * well. Depending on the current Database-engine, this can be done by using transaction or with table-locks
+ *
+ * @author Fabian Schmid <fs@studer-raimann.ch>
+ */
+class ilAtomQuery {
+
+	// Lock levels
+	const LOCK_WRITE = 1;
+	const LOCK_READ = 2;
+	// Isolation-Levels
+	const ISOLATION_READ_UNCOMMITED = 1;
+	const ISOLATION_READ_COMMITED = 2;
+	const ISOLATION_REPEATED_READ = 3;
+	const ISOLATION_SERIALIZABLE = 4;
+	/**
+	 * @var array
+	 */
+	protected static $available_isolations_levels = array(
+		self::ISOLATION_READ_UNCOMMITED,
+		self::ISOLATION_READ_COMMITED,
+		self::ISOLATION_REPEATED_READ,
+		self::ISOLATION_SERIALIZABLE,
+	);
+	// Anomalies
+	const ANO_LOST_UPDATES = 1;
+	const ANO_DIRTY_READ = 2;
+	const ANO_NON_REPEATED_READ = 3;
+	const ANO_PHANTOM = 4;
+	/**
+	 * @var array
+	 */
+	protected static $possible_anomalies = array(
+		self::ANO_LOST_UPDATES,
+		self::ANO_DIRTY_READ,
+		self::ANO_NON_REPEATED_READ,
+		self::ANO_PHANTOM,
+	);
+	/**
+	 * @var array
+	 */
+	protected static $anomalies_map = array(
+		self::ISOLATION_READ_UNCOMMITED => array(
+			self::ANO_LOST_UPDATES,
+			self::ANO_DIRTY_READ,
+			self::ANO_NON_REPEATED_READ,
+			self::ANO_PHANTOM,
+		),
+		self::ISOLATION_READ_COMMITED   => array(
+			self::ANO_NON_REPEATED_READ,
+			self::ANO_PHANTOM,
+		),
+		self::ISOLATION_REPEATED_READ   => array(
+			self::ANO_PHANTOM,
+		),
+		self::ISOLATION_SERIALIZABLE    => array(),
+	);
+	/**
+	 * @var int
+	 */
+	protected $isolation_level = self::ISOLATION_SERIALIZABLE;
+	/**
+	 * @var array
+	 */
+	protected $tables = array();
+	/**
+	 * @var callable
+	 */
+	protected $query;
+	/**
+	 * @var Closure[]
+	 */
+	protected $queries = array();
+	/**
+	 * @var
+	 */
+	protected $ilDBInstance;
+	/**
+	 * @var ilAtomQuery
+	 */
+	protected static $instance;
+
+
+	/**
+	 * ilAtomQuery constructor.
+	 *
+	 * @param \ilDBInterface $ilDBInstance
+	 * @param int $isolation_level currently only ISOLATION_SERIALIZABLE is available
+	 */
+	public function __construct(ilDBInterface $ilDBInstance, $isolation_level = self::ISOLATION_SERIALIZABLE) {
+		$this->ilDBInstance = $ilDBInstance;
+		$this->isolation_level = $isolation_level;
+	}
+
+	//
+	//
+	//
+	/**
+	 * @return array
+	 */
+	public function getRisks() {
+		return static::getPossibleAnomalies($this->getIsolationLevel());
+	}
+
+
+	/**
+	 * Add table-names which are influenced by your queries, MyISAm has to lock those tables. Lock
+	 *
+	 * @param string $table_name
+	 * @param int $lock_level use ilAtomQuery::LOCK_READ or ilAtomQuery::LOCK_WRITE
```

Comment:
currently only ISOLATION_SERIALIZABLE is possible, which is also needed when firing LOCK_WRITEs